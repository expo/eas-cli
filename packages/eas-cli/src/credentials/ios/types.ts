import { JSONObject } from '@expo/json-file';
import type { XCBuildConfiguration } from 'xcode';

import {
  AccountFragment,
  CommonIosAppCredentialsFragment,
  IosAppBuildCredentialsFragment,
} from '../../graphql/generated';

export interface App {
  account: AccountFragment;
  projectName: string;
}

export interface Target {
  targetName: string;
  buildConfiguration?: string;
  bundleIdentifier: string;
  parentBundleIdentifier?: string;
  entitlements: JSONObject;
  buildSettings?: XCBuildConfiguration['buildSettings'];
}

export interface TargetCredentials {
  distributionCertificate: {
    certificateP12: string;
    certificatePassword: string;
  };
  provisioningProfile: string;
}

export type IosCredentials = Record<string, TargetCredentials>;

export type IosAppBuildCredentialsMap = Record<string, IosAppBuildCredentialsFragment>;
export type IosAppCredentialsMap = Record<string, CommonIosAppCredentialsFragment | null>;
