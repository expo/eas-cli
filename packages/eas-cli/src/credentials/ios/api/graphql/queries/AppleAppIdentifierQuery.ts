import { print } from 'graphql';
import gql from 'graphql-tag';

import { ExpoGraphqlClient } from '../../../../../commandUtils/context/contextUtils/createGraphqlClient';
import { withErrorHandlingAsync } from '../../../../../graphql/client';
import {
  AppleAppIdentifierByBundleIdQuery,
  AppleAppIdentifierFragment,
} from '../../../../../graphql/generated';
import { AppleAppIdentifierFragmentNode } from '../../../../../graphql/types/credentials/AppleAppIdentifier';

export const AppleAppIdentifierQuery = {
  async byBundleIdentifierAsync(
    graphqlClient: ExpoGraphqlClient,
    accountName: string,
    bundleIdentifier: string
  ): Promise<AppleAppIdentifierFragment | null> {
    const data = await withErrorHandlingAsync(
      graphqlClient
        .query<AppleAppIdentifierByBundleIdQuery>(
          gql`
            query AppleAppIdentifierByBundleIdQuery(
              $accountName: String!
              $bundleIdentifier: String!
            ) {
              account {
                byName(accountName: $accountName) {
                  id
                  appleAppIdentifiers(bundleIdentifier: $bundleIdentifier) {
                    id
                    ...AppleAppIdentifierFragment
                  }
                }
              }
            }
            ${print(AppleAppIdentifierFragmentNode)}
          `,
          {
            accountName,
            bundleIdentifier,
          },
          {
            additionalTypenames: ['AppleAppIdentifier'],
          }
        )
        .toPromise()
    );
    return data.account.byName.appleAppIdentifiers[0];
  },
};
