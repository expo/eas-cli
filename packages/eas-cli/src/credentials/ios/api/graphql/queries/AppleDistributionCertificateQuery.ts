import assert from 'assert';
import { print } from 'graphql';
import gql from 'graphql-tag';

import { graphqlClient, withErrorHandlingAsync } from '../../../../../graphql/client';
import {
  AppleDistributionCertificateByAccountQuery,
  AppleDistributionCertificateByAppQuery,
  AppleDistributionCertificateFragment,
  IosDistributionType,
} from '../../../../../graphql/generated';
import { AppleDistributionCertificateFragmentNode } from '../../../../../graphql/types/credentials/AppleDistributionCertificate';
import { AppleTeamFragmentNode } from '../../../../../graphql/types/credentials/AppleTeam';

const AppleDistributionCertificateQuery = {
  async getForAppAsync(
    projectFullName: string,
    {
      appleAppIdentifierId,
      iosDistributionType,
    }: { appleAppIdentifierId: string; iosDistributionType: IosDistributionType }
  ): Promise<AppleDistributionCertificateFragment | null> {
    const data = await withErrorHandlingAsync(
      graphqlClient
        .query<AppleDistributionCertificateByAppQuery>(
          gql`
            query AppleDistributionCertificateByAppQuery(
              $projectFullName: String!
              $appleAppIdentifierId: String!
              $iosDistributionType: IosDistributionType!
            ) {
              app {
                byFullName(fullName: $projectFullName) {
                  id
                  iosAppCredentials(filter: { appleAppIdentifierId: $appleAppIdentifierId }) {
                    id
                    iosAppBuildCredentialsArray(
                      filter: { iosDistributionType: $iosDistributionType }
                    ) {
                      id
                      distributionCertificate {
                        id
                        ...AppleDistributionCertificateFragment
                        appleTeam {
                          id
                          ...AppleTeamFragment
                        }
                      }
                    }
                  }
                }
              }
            }
            ${print(AppleDistributionCertificateFragmentNode)}
            ${print(AppleTeamFragmentNode)}
          `,
          {
            projectFullName,
            appleAppIdentifierId,
            iosDistributionType,
          },
          {
            additionalTypenames: ['IosAppCredentials', 'IosAppBuildCredentials'],
          }
        )
        .toPromise()
    );
    assert(data.app, 'GraphQL: `app` not defined in server response');
    return (
      data.app.byFullName.iosAppCredentials[0]?.iosAppBuildCredentialsArray[0]
        ?.distributionCertificate ?? null
    );
  },
  async getAllForAccount(accountName: string): Promise<AppleDistributionCertificateFragment[]> {
    const data = await withErrorHandlingAsync(
      graphqlClient
        .query<AppleDistributionCertificateByAccountQuery>(
          gql`
            query AppleDistributionCertificateByAccountQuery($accountName: String!) {
              account {
                byName(accountName: $accountName) {
                  id
                  appleDistributionCertificates {
                    id
                    ...AppleDistributionCertificateFragment
                  }
                }
              }
            }
            ${print(AppleDistributionCertificateFragmentNode)}
          `,
          {
            accountName,
          }
        )
        .toPromise()
    );
    return data.account.byName.appleDistributionCertificates;
  },
};

export { AppleDistributionCertificateQuery };
