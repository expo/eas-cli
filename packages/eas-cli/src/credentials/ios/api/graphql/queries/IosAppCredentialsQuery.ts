import assert from 'assert';
import { print } from 'graphql';
import gql from 'graphql-tag';

import { ExpoGraphqlClient } from '../../../../../commandUtils/context/contextUtils/createGraphqlClient';
import { withErrorHandlingAsync } from '../../../../../graphql/client';
import {
  CommonIosAppCredentialsFragment,
  CommonIosAppCredentialsWithBuildCredentialsByAppIdentifierIdQuery,
  IosAppCredentialsWithBuildCredentialsByAppIdentifierIdQuery,
  IosDistributionType,
} from '../../../../../graphql/generated';
import { IosAppBuildCredentialsFragmentNode } from '../../../../../graphql/types/credentials/IosAppBuildCredentials';
import {
  CommonIosAppCredentialsFragmentNode,
  CommonIosAppCredentialsWithoutBuildCredentialsFragmentNode,
} from '../../../../../graphql/types/credentials/IosAppCredentials';

export const IosAppCredentialsQuery = {
  async withBuildCredentialsByAppIdentifierIdAsync(
    graphqlClient: ExpoGraphqlClient,
    projectFullName: string,
    {
      appleAppIdentifierId,
      iosDistributionType,
    }: {
      appleAppIdentifierId: string;
      iosDistributionType?: IosDistributionType;
    }
  ): Promise<CommonIosAppCredentialsFragment | null> {
    const data = await withErrorHandlingAsync(
      graphqlClient
        .query<IosAppCredentialsWithBuildCredentialsByAppIdentifierIdQuery>(
          gql`
            query IosAppCredentialsWithBuildCredentialsByAppIdentifierIdQuery(
              $projectFullName: String!
              $appleAppIdentifierId: String!
              $iosDistributionType: IosDistributionType
            ) {
              app {
                byFullName(fullName: $projectFullName) {
                  id
                  iosAppCredentials(filter: { appleAppIdentifierId: $appleAppIdentifierId }) {
                    id
                    ...CommonIosAppCredentialsWithoutBuildCredentialsFragment
                    iosAppBuildCredentialsList(
                      filter: { iosDistributionType: $iosDistributionType }
                    ) {
                      id
                      ...IosAppBuildCredentialsFragment
                    }
                  }
                }
              }
            }
            ${print(CommonIosAppCredentialsWithoutBuildCredentialsFragmentNode)}
            ${print(IosAppBuildCredentialsFragmentNode)}
          `,
          {
            projectFullName,
            appleAppIdentifierId,
            iosDistributionType,
          },
          {
            additionalTypenames: ['IosAppCredentials', 'IosAppBuildCredentials'],
          }
        )
        .toPromise()
    );
    assert(data.app, 'GraphQL: `app` not defined in server response');
    return data.app.byFullName.iosAppCredentials[0] ?? null;
  },
  async withCommonFieldsByAppIdentifierIdAsync(
    graphqlClient: ExpoGraphqlClient,
    projectFullName: string,
    {
      appleAppIdentifierId,
    }: {
      appleAppIdentifierId: string;
    }
  ): Promise<CommonIosAppCredentialsFragment | null> {
    const data = await withErrorHandlingAsync(
      graphqlClient
        .query<CommonIosAppCredentialsWithBuildCredentialsByAppIdentifierIdQuery>(
          gql`
            query CommonIosAppCredentialsWithBuildCredentialsByAppIdentifierIdQuery(
              $projectFullName: String!
              $appleAppIdentifierId: String!
            ) {
              app {
                byFullName(fullName: $projectFullName) {
                  id
                  iosAppCredentials(filter: { appleAppIdentifierId: $appleAppIdentifierId }) {
                    id
                    ...CommonIosAppCredentialsFragment
                  }
                }
              }
            }
            ${CommonIosAppCredentialsFragmentNode}
          `,
          {
            projectFullName,
            appleAppIdentifierId,
          },
          {
            additionalTypenames: ['IosAppCredentials', 'IosAppBuildCredentials', 'App'],
          }
        )
        .toPromise()
    );
    assert(data.app, 'GraphQL: `app` not defined in server response');
    return data.app.byFullName.iosAppCredentials[0] ?? null;
  },
};
