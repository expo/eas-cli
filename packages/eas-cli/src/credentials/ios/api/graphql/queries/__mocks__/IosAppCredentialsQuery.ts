export const IosAppCredentialsQuery = {
  withCommonFieldsByAppIdentifierIdAsync: jest.fn().mockImplementation(() => {
    return {
      id: '48d0cd34-96d8-447c-9de8-40dd2c3ca0f8',
      app: {
        id: '033134f3-e85c-4bde-9065-41a9c066ec31',
        fullName: '@quinlanj/test52',
        __typename: 'App',
      },
      appleTeam: {
        id: '6d39dbfe-a30a-44a4-bc36-d562e5963fa3',
        appleTeamIdentifier: 'test-apple-identifier',
        appleTeamName: null,
        __typename: 'AppleTeam',
      },
      appleAppIdentifier: {
        id: 'bd9c3609-26b9-451e-8ba6-e0604736c7fc',
        bundleIdentifier: 'com.quinlanj.test52',
        __typename: 'AppleAppIdentifier',
      },
      pushKey: {
        id: '74bdd007-da7b-4190-a571-edb87e9324a1',
        keyIdentifier: '9839M6AY8W',
        updatedAt: '2020-10-13T18:39:04.463Z',
        appleTeam: {
          id: '6d39dbfe-a30a-44a4-bc36-d562e5963fa3',
          appleTeamIdentifier: '77KQ969CHE',
          appleTeamName: null,
          __typename: 'AppleTeam',
        },
        __typename: 'ApplePushKey',
      },
      appStoreConnectApiKeyForSubmissions: {
        id: 'b8fe2225-3024-4ebe-9e36-7b9022d3b13e',
        appleTeam: {
          id: 'ca27f9f6-c7cb-40d1-8c41-c168e8548261',
          appleTeamIdentifier: '77KQ969CHE',
          appleTeamName: null,
          __typename: 'AppleTeam',
        },
        issuerIdentifier: '69a6de93-e4e4-47e3-e053-5b8c7c11a4d1',
        keyIdentifier: 'NYN468XNN3',
        name: 'Expo Submissions pl94k8_lnS',
        roles: ['ADMIN'],
        updatedAt: '2020-10-13T18:39:04.463Z',
        __typename: 'AppStoreConnectApiKey',
      },
      iosAppBuildCredentialsList: [
        {
          id: '6f8d0175-0caf-4922-b09b-eae757b83144',
          iosDistributionType: 'APP_STORE',
          distributionCertificate: {
            id: '8b3baa53-b149-4a1b-842b-75ff6e0bcb1e',
            certificateP12: 'test-cert-p12',
            certificatePassword: 'test-cert-password',
            serialNumber: 'test-serial',
            developerPortalIdentifier: 'test-portal-id',
            validityNotBefore: '2020-10-13T18:28:44.000Z',
            validityNotAfter: '2021-10-13T18:28:44.000Z',
            updatedAt: '2020-10-13T18:38:45.275Z',
            appleTeam: {
              id: '6d39dbfe-a30a-44a4-bc36-d562e5963fa3',
              appleTeamIdentifier: 'test-apple-identifier',
              appleTeamName: null,
              __typename: 'AppleTeam',
            },
            __typename: 'AppleDistributionCertificate',
          },
          provisioningProfile: {
            id: '01517f14-b4cf-489e-b1fd-3f5a61cb8df8',
            expiration: '2021-10-13T18:28:44.000Z',
            developerPortalIdentifier: 'test-profile-id',
            provisioningProfile: 'test-profile',
            updatedAt: '2020-10-29T00:18:32.130Z',
            status: 'active',
            appleDevices: [],
            appleTeam: {
              id: '6d39dbfe-a30a-44a4-bc36-d562e5963fa3',
              appleTeamIdentifier: 'test-apple-identifier',
              appleTeamName: null,
              __typename: 'AppleTeam',
            },
            __typename: 'AppleProvisioningProfile',
          },
          __typename: 'IosAppBuildCredentials',
        },
        {
          id: '3e1907a8-7114-4f14-9964-73467bc010ea',
          iosDistributionType: 'AD_HOC',
          distributionCertificate: {
            id: '8b3baa53-b149-4a1b-842b-75ff6e0bcb1e',
            certificateP12: 'test-cert-p12',
            certificatePassword: 'test-cert-password',
            serialNumber: 'test-serial',
            developerPortalIdentifier: 'test-portal-id',
            validityNotBefore: '2020-10-13T18:28:44.000Z',
            validityNotAfter: '2021-10-13T18:28:44.000Z',
            updatedAt: '2020-10-13T18:38:45.275Z',
            appleTeam: {
              id: '6d39dbfe-a30a-44a4-bc36-d562e5963fa3',
              appleTeamIdentifier: 'test-apple-identifier',
              appleTeamName: null,
              __typename: 'AppleTeam',
            },
            __typename: 'AppleDistributionCertificate',
          },
          provisioningProfile: {
            id: '01517f14-b4cf-489e-b1fd-3f5a61cb8df8',
            expiration: '2021-10-13T18:28:44.000Z',
            developerPortalIdentifier: 'test-profile-id',
            provisioningProfile: 'test-profile',
            updatedAt: '2020-10-29T00:18:32.130Z',
            status: 'active',
            appleDevices: [],
            appleTeam: {
              id: '6d39dbfe-a30a-44a4-bc36-d562e5963fa3',
              appleTeamIdentifier: 'test-apple-identifier',
              appleTeamName: null,
              __typename: 'AppleTeam',
            },
            __typename: 'AppleProvisioningProfile',
          },
          __typename: 'IosAppBuildCredentials',
        },
      ],
      __typename: 'IosAppCredentials',
    };
  }),
};
