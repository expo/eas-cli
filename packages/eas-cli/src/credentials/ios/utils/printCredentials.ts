import chalk from 'chalk';
import dateformat from 'dateformat';

import {
  AppleAppSpecificPasswordFragment,
  AppleDeviceFragment,
  ApplePushKeyFragment,
  IosAppBuildCredentialsFragment,
  IosDistributionType,
} from '../../../graphql/generated';
import { APPLE_DEVICE_CLASS_LABELS } from '../../../graphql/types/credentials/AppleDevice';
import Log from '../../../log';
import { fromNow } from '../../../utils/date';
import { AppLookupParams } from '../api/GraphqlClient';
import { App, IosAppBuildCredentialsMap, IosAppCredentialsMap, Target } from '../types';

function prettyIosDistributionType(distributionType: IosDistributionType): string {
  switch (distributionType) {
    case IosDistributionType.AppStore:
      return 'App Store';
    case IosDistributionType.AdHoc:
      return 'Ad Hoc';
    case IosDistributionType.Development:
      return 'Development';
    case IosDistributionType.Enterprise:
      return 'Enterprise';
    default:
      return 'Unknown';
  }
}

export function displayEmptyIosCredentials(appLookupParams: AppLookupParams): void {
  const { projectName, bundleIdentifier } = appLookupParams;
  Log.log(chalk.bold(`iOS Credentials`));
  Log.log(`  Project: ${projectName}`);
  Log.log(`  Bundle Identifier: ${bundleIdentifier}`);
  Log.log(`  No credentials set up yet!`);
}

/**
 * sort a build credentials array in descending order of preference
 */
function sortBuildCredentialsByDistributionType(
  iosAppBuildCredentialsList: IosAppBuildCredentialsFragment[]
): IosAppBuildCredentialsFragment[] {
  // The order in which we choose the distribution type from least to most preferred
  const typePriority = [
    IosDistributionType.Development,
    IosDistributionType.AdHoc,
    IosDistributionType.Enterprise,
    IosDistributionType.AppStore,
  ];
  return iosAppBuildCredentialsList
    .sort(
      (buildCredentialsA, buildCredentialsB) =>
        typePriority.indexOf(buildCredentialsA.iosDistributionType) -
        typePriority.indexOf(buildCredentialsB.iosDistributionType)
    )
    .reverse();
}

export function displayIosCredentials(
  app: App,
  appCredentialsMap: IosAppCredentialsMap,
  targets: Target[]
): void {
  const projectFullName = `@${app.account.name}/${app.projectName}`;
  const isMultitarget = targets.length > 1;

  Log.log(chalk.bold(`iOS Credentials`));
  Log.addNewLineIfNone();
  Log.log(`  Project: ${chalk.bold(projectFullName)}`);

  for (const { targetName, bundleIdentifier } of targets) {
    if (isMultitarget) {
      Log.newLine();
      Log.log(`  Target: ${chalk.bold(targetName)}`);
    }
    Log.log(`  Bundle Identifier: ${chalk.bold(bundleIdentifier)}`);
    const targetAppCredentials = appCredentialsMap[targetName];
    if (!targetAppCredentials) {
      Log.newLine();
      Log.log(`  No credentials set up yet!`);
      Log.newLine();
      continue;
    }

    const { appleTeam, pushKey, appSpecificPassword } = targetAppCredentials;
    if (appleTeam) {
      const { appleTeamIdentifier, appleTeamName } = appleTeam;
      Log.log(`  Apple Team: ${appleTeamIdentifier} ${appleTeamName ? `(${appleTeamName})` : ''}`);
    }
    Log.newLine();

    if (pushKey) {
      displayApplePushKey(pushKey);
    }

    if (appSpecificPassword) {
      displayAppSpecificPassword(appSpecificPassword);
    }

    const sortedIosAppBuildCredentialsList = sortBuildCredentialsByDistributionType(
      targetAppCredentials.iosAppBuildCredentialsList
    );
    for (const iosAppBuildCredentials of sortedIosAppBuildCredentialsList) {
      displayIosAppBuildCredentials(iosAppBuildCredentials);
    }
  }
}

export function displayProjectCredentials(
  app: App,
  appBuildCredentials: IosAppBuildCredentialsMap,
  targets: Target[]
): void {
  const projectFullName = `@${app.account.name}/${app.projectName}`;
  const targetToBundleId = targets.reduce<Record<string, string>>((acc, target) => {
    acc[target.targetName] = target.bundleIdentifier;
    return acc;
  }, {});
  const isMultitarget = targets.length > 1;

  Log.addNewLineIfNone();
  Log.log(chalk.bold('Project Credentials Configuration:'));
  Log.log(`  Project: ${chalk.bold(projectFullName)}`);
  for (const [targetName, buildCredentials] of Object.entries(appBuildCredentials)) {
    if (isMultitarget) {
      Log.newLine();
      Log.log(`  Target: ${chalk.bold(targetName)}`);
    }
    Log.log(`  Bundle Identifier: ${chalk.bold(targetToBundleId[targetName])}`);
    displayIosAppBuildCredentials(buildCredentials);
  }
}

function displayIosAppBuildCredentials(buildCredentials: IosAppBuildCredentialsFragment): void {
  Log.log(
    `  Configuration: ${chalk.bold(
      prettyIosDistributionType(buildCredentials.iosDistributionType)
    )}`
  );
  Log.newLine();
  const maybeDistCert = buildCredentials.distributionCertificate;
  Log.log(`  Distribution Certificate:`);
  if (maybeDistCert) {
    const { serialNumber, updatedAt, validityNotAfter, appleTeam } = maybeDistCert;
    Log.log(`    Serial Number: ${serialNumber}`);
    Log.log(`    Expiration Date: ${dateformat(validityNotAfter, 'expiresHeaderFormat')}`);
    if (appleTeam) {
      const { appleTeamIdentifier, appleTeamName } = appleTeam;
      Log.log(
        `    Apple Team: ${appleTeamIdentifier} ${appleTeamName ? `(${appleTeamName})` : ''}`
      );
    }
    Log.log(`    Updated ${fromNow(new Date(updatedAt))} ago`);
  } else {
    Log.log(`    None assigned yet`);
  }
  Log.newLine();

  const maybeProvProf = buildCredentials.provisioningProfile;
  Log.log(`  Provisioning Profile:`);
  if (maybeProvProf) {
    const { expiration, updatedAt, status, developerPortalIdentifier, appleTeam, appleDevices } =
      maybeProvProf;
    if (developerPortalIdentifier) {
      Log.log(`    Developer Portal ID: ${developerPortalIdentifier}`);
    }
    Log.log(`    Status: ${status}`);
    Log.log(`    Expiration Date: ${dateformat(expiration, 'expiresHeaderFormat')}`);
    if (appleTeam) {
      const { appleTeamIdentifier, appleTeamName } = appleTeam;
      Log.log(
        `    Apple Team: ${appleTeamIdentifier} ${appleTeamName ? `(${appleTeamName})` : ''}`
      );
    }
    if (appleDevices && appleDevices.length > 0) {
      Log.log(`    Provisioned devices:`);
      for (const appleDevice of appleDevices) {
        Log.log(`    - ${formatAppleDevice(appleDevice)}`);
      }
    }
    Log.log(`    Updated ${fromNow(new Date(updatedAt))} ago`);
  } else {
    Log.log(`    None assigned yet`);
  }
  Log.newLine();
}

function displayAppSpecificPassword(
  maybeAppSpecificPassword: AppleAppSpecificPasswordFragment | null
): void {
  Log.log(`  App Specific Password:`);
  if (maybeAppSpecificPassword) {
    const { appleIdUsername, passwordLabel, updatedAt } = maybeAppSpecificPassword;
    Log.log(`    Apple ID Username: ${appleIdUsername}`);
    if (passwordLabel) {
      Log.log(`    Password Label: ${passwordLabel}`);
    }
    Log.log(`    Updated ${fromNow(new Date(updatedAt))} ago`);
  } else {
    Log.log(`    None assigned yet`);
  }
  Log.newLine();
}

function displayApplePushKey(maybePushKey: ApplePushKeyFragment | null): void {
  Log.log(`  Push Key:`);
  if (maybePushKey) {
    const { keyIdentifier, appleTeam, updatedAt } = maybePushKey;
    Log.log(`    Developer Portal ID: ${keyIdentifier}`);
    if (appleTeam) {
      const { appleTeamIdentifier, appleTeamName } = appleTeam;
      Log.log(
        `    Apple Team: ${appleTeamIdentifier} ${appleTeamName ? `(${appleTeamName})` : ''}`
      );
    }
    Log.log(`    Updated ${fromNow(new Date(updatedAt))} ago`);
  } else {
    Log.log(`    None assigned yet`);
  }
  Log.newLine();
}

function formatAppleDevice(device: AppleDeviceFragment): string {
  let deviceString = '';
  if (device.name) {
    deviceString += device.name;
  }
  if (device.deviceClass || device.model) {
    const deviceDetails = [
      device.deviceClass && APPLE_DEVICE_CLASS_LABELS[device.deviceClass],
      device.model,
    ]
      .filter(i => i)
      .join(' ');
    if (deviceString === '') {
      deviceString += deviceDetails;
    } else {
      deviceString += ` - ${deviceDetails}`;
    }
  }

  if (deviceString === '') {
    deviceString += device.identifier;
  } else {
    deviceString += ` (UDID: ${device.identifier})`;
  }
  return deviceString;
}
