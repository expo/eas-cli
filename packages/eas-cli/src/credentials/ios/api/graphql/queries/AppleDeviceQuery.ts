import assert from 'assert';
import { print } from 'graphql';
import gql from 'graphql-tag';

import { ExpoGraphqlClient } from '../../../../../commandUtils/context/contextUtils/createGraphqlClient';
import { DeviceNotFoundError } from '../../../../../devices/utils/errors';
import { withErrorHandlingAsync } from '../../../../../graphql/client';
import {
  AppleDevice,
  AppleDeviceFragment,
  AppleDevicesByAppleTeamQuery,
  AppleDevicesByIdentifierQuery,
  AppleDevicesByTeamIdentifierQuery,
  AppleDevicesByTeamIdentifierQueryVariables,
  AppleTeamFragment,
} from '../../../../../graphql/generated';
import { AppleDeviceFragmentNode } from '../../../../../graphql/types/credentials/AppleDevice';
import { AppleTeamFragmentNode } from '../../../../../graphql/types/credentials/AppleTeam';
import { formatAppleTeam } from '../../../actions/AppleTeamFormatting';

export type AppleDeviceFragmentWithAppleTeam = AppleDeviceFragment & {
  appleTeam: AppleTeamFragment;
};

export type AppleDeviceQueryResult = Pick<
  AppleDevice,
  'id' | 'identifier' | 'name' | 'deviceClass' | 'enabled'
>;

export type AppleDevicesByTeamIdentifierQueryResult = AppleTeamFragment & {
  appleDevices: AppleDeviceQueryResult[];
};

export type AppleDevicesByIdentifierQueryResult = AppleDeviceQueryResult & {
  appleTeam: AppleTeamFragment;
};

export const AppleDeviceQuery = {
  async getAllByAppleTeamIdentifierAsync(
    graphqlClient: ExpoGraphqlClient,
    accountId: string,
    appleTeamIdentifier: string,
    { useCache = true }: { useCache?: boolean } = {}
  ): Promise<AppleDeviceFragmentWithAppleTeam[]> {
    const data = await withErrorHandlingAsync(
      graphqlClient
        .query<AppleDevicesByAppleTeamQuery>(
          gql`
            query AppleDevicesByAppleTeamQuery($accountId: ID!, $appleTeamIdentifier: String!) {
              appleTeam {
                byAppleTeamIdentifier(accountId: $accountId, identifier: $appleTeamIdentifier) {
                  id
                  ...AppleTeamFragment
                  appleDevices {
                    id
                    ...AppleDeviceFragment
                    appleTeam {
                      id
                      ...AppleTeamFragment
                    }
                  }
                }
              }
            }
            ${print(AppleTeamFragmentNode)}
            ${print(AppleDeviceFragmentNode)}
          `,
          {
            accountId,
            appleTeamIdentifier,
          },
          {
            additionalTypenames: ['AppleDevice'],
            requestPolicy: useCache ? 'cache-first' : 'network-only',
          }
        )
        .toPromise()
    );
    assert(
      data.appleTeam.byAppleTeamIdentifier,
      'byAppleTeamIdentifier should be defined in this context - enforced by GraphQL'
    );
    const { appleDevices } = data.appleTeam.byAppleTeamIdentifier;
    assert(appleDevices, 'Apple Devices should be defined in this context - enforced by GraphQL');
    return appleDevices;
  },

  async getAllForAppleTeamAsync(
    graphqlClient: ExpoGraphqlClient,
    { accountName, appleTeamIdentifier, offset, limit }: AppleDevicesByTeamIdentifierQueryVariables
  ): Promise<AppleDeviceFragment[]> {
    const data = await withErrorHandlingAsync(
      graphqlClient
        .query<AppleDevicesByTeamIdentifierQuery>(
          gql`
            query AppleDevicesByTeamIdentifier(
              $accountName: String!
              $appleTeamIdentifier: String!
              $offset: Int
              $limit: Int
            ) {
              account {
                byName(accountName: $accountName) {
                  id
                  appleTeams(appleTeamIdentifier: $appleTeamIdentifier) {
                    id
                    appleTeamIdentifier
                    appleTeamName
                    appleDevices(offset: $offset, limit: $limit) {
                      id
                      identifier
                      name
                      deviceClass
                      enabled
                      model
                    }
                  }
                }
              }
            }
          `,
          { accountName, appleTeamIdentifier, offset, limit },
          {
            additionalTypenames: ['AppleDevice', 'AppleTeam'],
          }
        )
        .toPromise()
    );
    const [appleTeam] = data.account.byName.appleTeams;
    const { appleDevices } = appleTeam;
    if (!appleDevices) {
      throw new Error(`Could not find devices on Apple team -- ${formatAppleTeam(appleTeam)}`);
    }
    return appleDevices;
  },

  async getByDeviceIdentifierAsync(
    graphqlClient: ExpoGraphqlClient,
    accountName: string,
    identifier: string
  ): Promise<AppleDevicesByIdentifierQueryResult> {
    const data = await withErrorHandlingAsync(
      graphqlClient
        .query<AppleDevicesByIdentifierQuery>(
          gql`
            query AppleDevicesByIdentifier($accountName: String!, $identifier: String!) {
              account {
                byName(accountName: $accountName) {
                  id
                  appleDevices(identifier: $identifier) {
                    id
                    model
                    identifier
                    name
                    deviceClass
                    enabled
                    appleTeam {
                      id
                      appleTeamIdentifier
                      appleTeamName
                    }
                  }
                }
              }
            }
          `,
          { accountName, identifier },
          {
            additionalTypenames: ['AppleDevice', 'AppleTeam'],
          }
        )
        .toPromise()
    );

    const device = data.account.byName.appleDevices[0];
    if (!device) {
      throw new DeviceNotFoundError(
        `Device with id ${identifier} was not found on account ${accountName}.`
      );
    }
    return device;
  },
};
