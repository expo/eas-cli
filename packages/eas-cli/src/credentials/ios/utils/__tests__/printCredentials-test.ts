import mockdate from 'mockdate';
import nullthrows from 'nullthrows';
import { instance, mock } from 'ts-mockito';

import { ExpoGraphqlClient } from '../../../../commandUtils/context/contextUtils/createGraphqlClient';
import Log from '../../../../log';
import { IosAppCredentialsQuery } from '../../api/graphql/queries/IosAppCredentialsQuery';
import { App, Target } from '../../types';
import { displayIosCredentials } from '../printCredentials';

jest.mock('../../../../log');
jest.mock('chalk', () => ({ bold: jest.fn(log => log), cyan: { bold: jest.fn(log => log) } }));

jest.mock('../../api/graphql/queries/IosAppCredentialsQuery');
mockdate.set(new Date('4/20/2021'));

describe('print credentials', () => {
  it('prints the IosAppCredentials map', async () => {
    const graphqlClient = instance(mock<ExpoGraphqlClient>());
    const app: App = {
      account: {
        id: 'account-id',
        name: 'quinlanj',
        users: [],
      },
      projectName: 'test52',
    };
    const testIosAppCredentialsData =
      await IosAppCredentialsQuery.withCommonFieldsByAppIdentifierIdAsync(
        graphqlClient,
        '@quinlanj/test52',
        {
          appleAppIdentifierId: 'test-id',
        }
      );
    const appCredentials = {
      test52: nullthrows(testIosAppCredentialsData),
    };
    const targets: Target[] = [
      { targetName: 'test52', bundleIdentifier: 'com.quinlanj.test52', entitlements: {} },
    ];
    displayIosCredentials(app, appCredentials, targets);
    const loggedSoFar = jest
      .mocked(Log.log)
      .mock.calls.reduce((acc, mockValue) => acc + mockValue, '');
    expect(loggedSoFar).toMatchSnapshot();
  });
});
