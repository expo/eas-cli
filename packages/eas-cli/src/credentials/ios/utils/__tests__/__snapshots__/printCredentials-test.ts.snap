// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`print credentials prints the IosAppCredentials map 1`] = `
"iOS Credentials            
Project                    @quinlanj/test52
Bundle Identifier          com.quinlanj.test52
Apple Team                 test-apple-identifier 
                           
Push Key                   
Developer Portal ID        9839M6AY8W
Apple Team                 77KQ969CHE 
Updated                    6 months ago
                           
App Store Connect API Key  
Developer Portal ID        NYN468XNN3
Name                       Expo Submissions pl94k8_lnS
Issuer ID                  69a6de93-e4e4-47e3-e053-5b8c7c11a4d1
Roles                      ADMIN
Apple Team                 77KQ969CHE 
Updated                    6 months ago
                           
                           
App Store Configuration    
                           
Distribution Certificate   
Serial Number              test-serial
Expiration Date            Wed, 13 Oct 2021 18:28:44 UTC
Apple Team                 test-apple-identifier 
Updated                    6 months ago
                           
Provisioning Profile       
Developer Portal ID        test-profile-id
Status                     active
Expiration                 Wed, 13 Oct 2021 18:28:44 UTC
Apple Team                 test-apple-identifier 
Updated                    5 months ago
                           
                           
Ad Hoc Configuration       
                           
Distribution Certificate   
Serial Number              test-serial
Expiration Date            Wed, 13 Oct 2021 18:28:44 UTC
Apple Team                 test-apple-identifier 
Updated                    6 months ago
                           
Provisioning Profile       
Developer Portal ID        test-profile-id
Status                     active
Expiration                 Wed, 13 Oct 2021 18:28:44 UTC
Apple Team                 test-apple-identifier 
Updated                    5 months ago
                           "
`;
